Implement loading old megolm keys from SSSS

TobiasFella opened the pull request:
Some security checks are still missing, also a bunch of refactoring and cleanup

Best enjoyed with https://invent.kde.org/network/neochat/-/merge_requests/1111

TobiasFella committed:
Implement loading key backup from security phrase

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -58,6 +58,15 @@
+#include <openssl/hmac.h>
+#include <openssl/kdf.h>
+#include <openssl/rand.h>
+#include <openssl/sha.h>
```
Can we have all those OpenSSL wrappers in a separate h/cpp file pair? I'd rather we try hard to reduce the size of `connection.cpp` and `room.cpp`. These things are quite self-contained and don't need Quotient context.

Btw, did you check the Qt(6) API, if they already wrapped some of these, by chance? Just asking so that I don't re-do what you likely already did.

KitsuneRal reviewed (changes requested):
Haven't reviewed the code in detail yet but got one thing to begin with, please.

TobiasFella committed:
Fix

TobiasFella committed:
Refactor

TobiasFella committed:
Refactor

TobiasFella committed:
Fixes

TobiasFella committed:
Implement getting key from other device

TobiasFella committed:
Implement SSSS from security key

KitsuneRal commented on Quotient/e2ee/qolminboundsession.h:
```diff
@@ -49,8 +49,9 @@ class QUOTIENT_API QOlmInboundGroupSession
     const char* lastError() const;
 
-private:
+    //TODO: why?
```
Because so far all `QOlmInboundGroupSession` objects were supposed to be created by static factory methods. If that's no more the case, feel free to change it - but you don't seem to need opening it up so far?

TobiasFella replied:
I'm still only using a factory method; the code fails to compile when the constructor is private, though

KitsuneRal replied:
Well, the current `Room::addMegolmSessionFromBackup()` code obviously tries to default-construct `QOlmInboundGroupSession` each time you try to access the session via `operator[]`. As I'm already hacking on this PR, I think I can rewrite it in a more idiomatic C++, with iterators and so.

KitsuneRal commented on Quotient/e2ee/sssshandler.cpp:
```diff
@@ -0,0 +1,227 @@
+QByteArray SSSSHandler::decryptKey(const QString& name, const QByteArray& decryptionKey) const
+{
+    Q_ASSERT(m_connection);
+    const auto defaultKey = m_connection->accountData("m.secret_storage.default_key"_ls)->contentPart<QString>("key"_ls);
```
Would be great to have it as an event type, too (and btw you don't really need to have a separate class for it, just use `DEFINE_SIMPLE_EVENT` before the first usage. 

This is unsafe, by the way - `accountData` may return `nullptr`. `Connection::accountDataContent<EventT>()` would be much more practical but it doesn't exist yet.

KitsuneRal commented on Quotient/e2ee/sssshandler.cpp:
```diff
@@ -0,0 +1,227 @@
+    Q_ASSERT(m_connection);
+    Connection::UsersToDevicesToContent content;
+    const auto& requestId = m_connection->generateTxnId();
+    QJsonObject eventContent;
+    eventContent["action"_ls] = "request"_ls;
+    eventContent["name"_ls] = "m.megolm_backup.v1"_ls;
+    eventContent["request_id"_ls] = requestId;
+    eventContent["requesting_device_id"_ls] = m_connection->deviceId();
+    for (const auto& deviceId : m_connection->devicesForUser(m_connection->userId())) {
+        content[m_connection->userId()][deviceId] = eventContent;
+    }
+    m_connection->sendToDevices("m.secret.request"_ls, content);
```
Wrap it in a proper event type, please... I know it's more work :)

KitsuneRal replied:
Ah, sorry, no, it's `sendToDevices()`, not `sendToDevice()`.

TobiasFella committed:
Cache cross-signing and megolm keys locally

TobiasFella committed:
Create event type and fix

TobiasFella committed:
More fixes

TobiasFella commented:
I think this is ready from my side. What's "missing" for a complete SSSS implementation is the "other side" of the whole thing - uploading keys to the backup, etc. - I'll do that in a different merge request.

One thing I'm not sure about is the verified status of the megolm sessions - thing like the `is_verified` and `forwarded` fields of the keys from the backup. I fail to see how those would be useful and we thus ignore them entirely so far. When replacing an existing, verified session with a session from the backup (which we only do if the session from the backup has a lower first message index), we manually keep the verified status - This probably won't really happen in practice, though.

TobiasFella committed:
Ifdefs

TobiasFella committed:
Remove custom job implementations

TobiasFella committed:
Store etag

KitsuneRal committed:
Drop unneeded testcryptoutils.h

KitsuneRal committed:
Brush up #includes

TobiasFella committed:
Merge branch 'dev' into tobias/ssss

KitsuneRal committed:
Room::addMegolmSessionFromBackup(): use more idiomatic C++

Using an iterator instead of repeated hashmap lookups is not only
slightly more efficient but also allows to make QOlmInboundSession
constructor private again.

KitsuneRal committed:
pbkdf2HmacSha512: default keyLength to 32

We don't use any other anywhere atm.

KitsuneRal committed:
Clean up testcryptoutils.cpp

Mostly replacing the magic constant which is QByteArray of 32 zeroes
with a function that returns it.

KitsuneRal committed:
Connection::accountData(): allow to override the account data key

This will be used in a subsequent commit to retrieve an "event" (account
data are not quite full-fledged events but the JSON shape is the same)
whose key is dynamic even though the data type is the same.

KitsuneRal committed:
Clean up from code review of SSSSHandler

TobiasFella committed:
Fix and add test

KitsuneRal committed:
ConnectionEncryptionData: make mock look more like a full-blown login

Generate a random pickling key even in case of a mock, and clear
the database at each test run.

KitsuneRal committed:
Merge remote-tracking branch 'origin/dev' into kitsune/ssss

fishman commented:
@KitsuneRal what's missing to get this merged? As it stands libQuotient isn't really usable for any encrypted communication

KitsuneRal commented:
In short, I'm still not quite happy with the current code quality, e.g. some errors are not handled. Given that this is encryption-related code I prefer to be very cautious about it, especially since the plan is to add it to the stable (0.8.x) branch. Another thing that holds me back atm is that I'm away on vacation with no access to my development machine until the end of November. I suggest using this branch directly in the meantime.

TobiasFella committed:
Store master key

TobiasFella committed:
Improve error handling

TobiasFella commented:
I've started to add some error handling. It's not exactly amazing, as the functions sometimes return openssl error code, sometimes libolm error codes - and sometimes both. Since the consumers of the crypto functions can't do anything about the errors anyway, we might consider just printing the exact errors in the crypto functions themselves, where we know how to get the correct error codes, and not expose them in the API at all.

KitsuneRal commented:
That's pretty much exactly what I had on my mind: return errors wrapped into `Quotient::Expected` where one can reasonably expect some error handling to be possible/necessary on the client side, and just log other errors.

Commit:
Merge branch 'dev' into kitsune/ssss

KitsuneRal commented on Quotient/e2ee/cryptoutils.cpp:
```diff
@@ -0,0 +1,256 @@
+
+    auto status = olm_pk_encryption_set_recipient_key(context.get(), publicKey.data(), publicKey.size());
+    if (status == olm_error()) {
+        return olm_pk_encryption_last_error_code(context.get());
```
Took me sometime to ~~figure out~~ remember why MSVC fails to compile this.

Welcome to the small lovely world of implementation-defined underlying types. In a nutshell, because `enum OlmErrorCode` doesn't specify its underlying type GCC and Clang assume it to be unsigned while MSVC thinks it should be signed. None of the opinions is wrong, the standard doesn't mandate either. So, when you compile this code with MSVC the compiler doesn't want to convert `OlmErrorCode` to an unsigned type without an explicit cast.

Given that the whole `Expected<>` construct is repeated across `cryptoutils.*`,  I suggest to add a `using` declaration for it, in the following form:
```cplusplus
template <typename T>
using OlmExpected = Expected<T, std::underlying_type_t<OlmErrorCode>>;
```
and use it throghout the codebase. HTH.

KitsuneRal replied:
...and I just found that we already have `QOlmExpected`... so the name for the one in this PR should probably be something less confusing.

KitsuneRal replied:
...and looking further into the code, I get an uneasy feeling that nothing guarantees error values from Olm and OpenSSL from clashing - so I'm afraid we just cannot pack things like that. Also, as discussed earlier here, probably we have to go through this code and see which errors are even worth returning - I suppose most of OpenSSL errors are not really recoverable and/or are effectively the same from the caller's perspective.

KitsuneRal replied:
I made a shot at fixing this along the lines mentioned in the chat. Didn't really like the fact that my changes didn't break anything outside of `cryptoutilus.*` - might mean having to look through all call sites and see how errors are dealt with _there_. But maybe it's fine...

Commit:
cryptoutils.*: Use SslExpected and QOlmExpected

Along with a bit of reformatting.

KitsuneRal commented on Quotient/e2ee/cryptoutils.cpp:
```diff
@@ -0,0 +1,267 @@
+    }
+    EVP_PKEY_CTX_free(context);
+
+    if (outputLength != 64) {
```
In this case, I guess we could put a `Q_ASSERT` but otherwise return the buffer, or, harsher, just `std::terminate()`? I don't see how the calling code can recover from this.

TobiasFella replied:
The caller could just... cancel the crypto stuff they're doing and... continue running; in practice, there's no reason for this to happen other than a bug in libquotient (or openssl), so terminating would be the best thing to do IMO

KitsuneRal replied:
As far as I gather, it is not a given that the output length should be exactly 64 bytes for any inputs; for cryptoutils, I decided to still return an error code from cryptoutils (fortunately, OpenSSL's error code space _is_ extensible, unlike Olm's) and later see if more assertions/terminations should be introduced up the stack.

Commit:
Fixup hkdfSha256()

1. Replace the magic number returned when outputLength is not 64, with
   an assertion; it's never supposed to be different from 64. Maybe
   we should even `std::terminate` here?
2. Swallow the repetitive code into a macro and clear the SSL context
   in that macro - previously the code leaked the context every time
   it finished with an error. (Those errors are usually unrecoverable
   but it's good to manage resources carefully anyway.)

Commit:
Fixup on fixup :-\

The CI tests didn't reveal the false-negative blunder in Q_ASSERT
because CI builds stuff in Release mode (which might or might not be
right but that's a matter for another PR), so it's replaced with
something more robust.

Commit:
cryptoutils: ContextHolder; drop extra status variables

Using ContextHolder plugs quite a few memory leaks on error conditions
occurring in cryptoutils. Most of those conditions are grave but that
doesn't mean we should go lax on resource management.

Commit:
cryptoutils: don't expose OpenSSL dependency

It's meant to be private, even if some constants and types happen to be
used in our interface.

Commit:
Replace RAND_bytes() with fillFromSecureRng(); FixedBuffer::operator[]

With this commit, fillFromSecureRng becomes a common way to fill up
a buffer with (truly, if possible) random data. QByteArray-friendly
overload can be made later if necessary.

Also: add operator[] to FixedBufferBase and FixedBuffer.

Commit:
cryptoutils: check sizes; constify where possible

checkedSize() makes sure that the passed 64-bit sizes can fit signed int
parameters commonly accepted throughout OpenSSL. When the size ends up
clamped, the crypto function that uses it will likely either fail (e.g.
in case of decryption) or act on the trimmed payload (encryption,
digests) but that will be the problem the caller will have to deal with.

Commit:
Bring static_cast<> back for the mask

Using small integer types inside the expression means integer promotion
rules start kicking in. In particular, and quite counterintuitively,
~(uint8_t{1}) has type int, not unsigned int. So instead, just do the
calculation with unsigned int and trim the result to uint8_t.

(thanks to Sonar for catching that)

Commit:
Cleanup

Mostly upon Sonar warnings

Commit:
More size checking; allow unsigned types in checkedSize()

KitsuneRal commented:
I ought to be more consistent with my identities... :)

Commit:
cryptoutils: more type safety with compile-time-sized spans

This commit introduces quite a bit stuff built around std::span to
make sure the data passed to OpenSSL and (eventually) to Olm use fewer
reinterpret_cast's, replacing them with easy to read and check calls
that, on top, work with compile-time-sized buffers and spans where
possible. At least some keys have moved to using secure memory along
the way - that still is far from being airtight on keeping sensitive
stuff in secure memory but already better than before.

Commit:
base58decode: constexpr-initialise reverse_alphabet

Commit:
More cleanup

KitsuneRal committed:
Code review: AesKeySize -> Aes256KeySize

KitsuneRal committed:
Update Quotient/e2ee/e2ee_common.h

Co-authored-by: Tobias Fella <9750016+TobiasFella@users.noreply.github.com>

Commit:
Fixes upon code review

Also: CLAMP_SIZE now needs a semicolon, like a "normal" statement.

Commit:
cryptoutils: replace remaining reinterpret_casts

Commit:
More validations, comments, reformatting

KitsuneRal committed:
Merge branch 'tobias/ssss' into kitsune/spans-all-over

KitsuneRal committed:
Merge pull request #716 from quotient-im/kitsune/spans-all-over

cryptoutils: more type safety with compile-time-sized spans

KitsuneRal commented on Quotient/e2ee/cryptoutils.cpp:
```diff
@@ -0,0 +1,315 @@
+
+    CLAMP_SIZE(ciphertextSize, ciphertext);
+
+    FixedBuffer<> decrypted(static_cast<size_t>(ciphertextSize),
+                            FixedBufferBase::FillWithZeros);
```
Not to myself who added this code in #716: this is not right. `FixedBuffer`s use secure memory, which is currently in a rather short supply (64 KiB) and only meant to store secrets. This code in particular is used to decrypt attachments, and a file of any considerable size now literally leads to an assertion failure/crash.

Commit:
Fix a regression in handling attachments

aesCtr256En/Decrypt used FixedBuffer for the target (encrypted and
decrypted, respectively) buffer. FixedBuffer uses secure memory, and
it is purposefully limited to 64k, because it's meant to be used for
secrets, not for payloads (for now at least). This commit reverts to
using normal QByteArray to store the result, using asWritableCBytes()
to get a span adapter that is more convenient to work with in OpenSSL.

Commit:
Fix include paths

...to work with QUOTIENT_FORCE_NAMESPACED_INCLUDES

Commit:
Merge branch 'dev' into kitsune/ssss

TobiasFella committed:
Add types for account data events

TobiasFella committed:
Improve error handling

Commit:
Merge branch 'dev' into ssss

Commit:
Namespace SSSSHandler; other refactoring

1. Instead of templating decryptKey (which effectively emits the same
   function 4 times), just pass the event type as a parameter.
2. Similarly, just use event type strings instead of full-blown Event
   derivatives.
3. requestKeyFromDevices() is pretty much all about doing stuff in
   the Connection context, so move it there.

KitsuneRal commented on Quotient/e2ee/sssshandler.cpp:
```diff
@@ -0,0 +1,260 @@
+
+void SSSSHandler::unlockSSSSFromCrossSigning()
+{
+    Q_ASSERT(m_connection);
```
Would it make sense to double-check the current device is verified, before even trying?

TobiasFella replied:
yes-ish: At the moment, we don't really know whether we're cross-signing verified (#630). We could limit the key requests to "one-to-one" verified devices, but i'd skip that for now and do it properly once #630 is merged

KitsuneRal replied:
Ah, right, got it.

Commit:
calculateDefaultKey() -> unlockAndLoad() + more refactoring

1. The rename is because it does so much more than just default key
   calculation.
2. Speaking of the default key, it's now only retrieved once, not 5
   times.
3. The default key "event" type definition can be removed now that
   usage of it becomes one-time access of a single key.

KitsuneRal commented on Quotient/e2ee/sssshandler.cpp:
```diff
@@ -0,0 +1,253 @@
+void SSSSHandler::unlockSSSSFromPassword(const QString& password)
+{
+    Q_ASSERT(m_connection);
+    unlockAndLoad(password.toLatin1(), true);
```
More of an aside comment for now (seems that [we have a gap in the spec](https://github.com/matrix-org/matrix-spec/issues/1750) here) but this code implies that the passphrase is in Latin-1 encoding. I guess we could assume that UTF-8 would use with other clients, too.

TobiasFella replied:
changed

KitsuneRal reviewed (commented):
A couple more things discovered while ~~messing with~~ refactoring the code.

TobiasFella committed:
Utf8 passphrase

KitsuneRal commented on Quotient/e2ee/cryptoutils.h:
```diff
@@ -0,0 +1,102 @@
+
+//! \brief Derive a key from the input data using HKDF-SHA256
+//!
+//! All input parameters have a length of 32 bytes
```
This seems to be not quite true, as `calculateDefaultKey()` (now dubbed `unlockAndLoad()`) passes an empty (so, zero-sized) `QByteArray` to it. I suppose it's the comment being inaccurate - or should the implementation pass 32 zero bytes instead?

KitsuneRal replied:
If I found the right place in the spec, [it says](https://spec.matrix.org/latest/client-server-api/#msecret_storagev1aes-hmac-sha2) we should be "using the empty string as the info parameter" - so it's the comment being inaccurate indeed.

KitsuneRal committed:
Refactor unlockAndLoad()

The method became rather large and with additional error checking would
get even larger - so now the initial preparations concerned with
default key and key description retrieval are put out to
UnlockData::prepare() - a static factory method returning the loaded
information that is used further in the process. That allows to put
the piece concerned with making the key out of the passphrase to
unlockSSSSWithPassphrase() (formerly not quite correctly named as
unlockSSSSFromPassword). More byte_views are introduced, to ensure
correct payload sizes as early (preferably at compile time) as possible.

KitsuneRal committed:
Use spans properly

Fixes building on macOS.

KitsuneRal committed:
loadMegolmBackup(): use facilities the jobs provide

Also: fix failing to jump over the result-using code in case of failing
to decrypt the session.

KitsuneRal commented:
This now looks ready from my side, too. Sorry for taking plenty of time to polish the code, I trust it was worth it though. Let me know if you're good to go with this.

KitsuneRal committed:
Cleanup

KitsuneRal committed:
Fix garbage in info parameter passed to hkdfSha256

Thanks to @tobias for spotting that.
